### PR TITLE
REST API: Support query by partial id

### PR DESCRIPTION
**Pull request by aboch**
- for networks and endpoints

Signed-off-by: Alessandro Boch <aboch@docker.com>

### CODE REVIEW

File: api/api.go [hunk 1]
```diff
@@ -234,12 +243,27 @@ func procGetNetwork(c libnetwork.NetworkController, vars map[string]string, body
...
-	if name, queryByName := vars[urlNwName]; queryByName {
-		nw, errRsp := findNetwork(c, name, byName)
-		if errRsp.isOK() {
+	// Look for query filters and validate
+	name, queryByName := vars[urlNwName]
+	shortID, queryByPid := vars[urlNwPID]
+	if queryByName && queryByPid {
+		// vars are populated from the URL, following cannot happen
+		panic("Unsupported query composition URL variable parameter for endpoint")
+	}
```
Comment (mavenugo) [inline api/api.go]: I dont think we should go to this extreme for testing purposes.
Please remove the above if condition

File: api/api.go [hunk 2]
```diff
@@ -234,12 +243,27 @@ func procGetNetwork(c libnetwork.NetworkController, vars map[string]string, body
...
+
+	if queryByName {
+		if nw, errRsp := findNetwork(c, name, byName); errRsp.isOK() {
 			list = append(list, buildNetworkResource(nw))
 		}
+	} else if queryByPid {
+		// Return all the prefix-matching networks
+		l := func(nw libnetwork.Network) bool {
+			if strings.HasPrefix(nw.ID(), shortID) {
```
Comment (mavenugo) [inline api/api.go]: Should we have to be case-sensitive ?
Just checking. Can you please verify docker behavior ?

File: api/api.go (hunk 1 above)
Comment (aboch) [inline api/api.go]: Even though this cannot happen (unless a bug in the mux), I will replace the panic() with a Forbidden response message.

File: api/api.go (hunk 2 above)
Comment (mavenugo) [inline api/api.go]: Confirmed in docker and it is case sensitive... all good.

**Comment by aboch** [comment]
@mavenugo Taken care of your comment. Please have a look when you get a chance.

**Comment by mavenugo** [comment]
Thanks. LGTM

**Comment by mrjana** [comment]
LGTM

**Pull request merged**